crates/provider/src/docker/namespace.rs

```diff
@@ -0,0 +1,356 @@
+    types::{
+        GenerateFileCommand, GenerateFilesOptions, ProviderCapabilities, RunCommandOptions,
+        SpawnNodeOptions,
+    },
+    DynNode, ProviderError, ProviderNamespace, ProviderNode,
+};
+
+use super::{
+    client::{ContainerRunOptions, DockerClient},
+    node::DockerNode,
+    DockerProvider,
+};
+
+pub struct DockerNamespace<FS>
+where
+    FS: FileSystem + Send + Sync + Clone,
+{
+    weak: Weak<DockerNamespace<FS>>,
+    #[allow(dead_code)]
+    provider: Weak<DockerProvider<FS>>,
+    name: String,
+    base_dir: PathBuf,
+    capabilities: ProviderCapabilities,
+    docker_client: DockerClient,
+    filesystem: FS,
+    delete_on_drop: Arc<Mutex<bool>>,
+    pub(super) nodes: RwLock<HashMap<String, Arc<DockerNode<FS>>>>,
+}
+
+impl<FS> DockerNamespace<FS>
+where
+    FS: FileSystem + Send + Sync + Clone + 'static,
+{
+    pub(super) async fn new(
+        provider: &Weak<DockerProvider<FS>>,
+        tmp_dir: &PathBuf,
+        capabilities: &ProviderCapabilities,
+        docker_client: &DockerClient,
+        filesystem: &FS,
+    ) -> Result<Arc<Self>, ProviderError> {
+        let name = format!("{}{}", NAMESPACE_PREFIX, Uuid::new_v4());
+        let base_dir = PathBuf::from_iter([tmp_dir, &PathBuf::from(&name)]);
+        filesystem.create_dir(&base_dir).await?;
+
+        let namespace = Arc::new_cyclic(|weak| DockerNamespace {
+            weak: weak.clone(),
+            provider: provider.clone(),
+            name,
+            base_dir,
+            capabilities: capabilities.clone(),
+            filesystem: filesystem.clone(),
+            docker_client: docker_client.clone(),
+            nodes: RwLock::new(HashMap::new()),
+            delete_on_drop: Arc::new(Mutex::new(true)),
+        });
+
+        namespace.initialize().await?;
+
+        Ok(namespace)
+    }
+
+    async fn initialize(&self) -> Result<(), ProviderError> {
+        self.initialize_zombie_wrapper_volume().await?;
+        self.initialize_helper_binaries_volume().await?;
+
+        Ok(())
+    }
+
+    async fn initialize_zombie_wrapper_volume(&self) -> Result<(), ProviderError> {
+        let local_zombie_wrapper_path =
+            PathBuf::from_iter([&self.base_dir, &PathBuf::from("zombie-wrapper.sh")]);
+
+        self.filesystem
+            .write(
+                &local_zombie_wrapper_path,
+                include_str!("../shared/scripts/zombie-wrapper.sh"),
+            )
+            .await?;
+
+        let zombie_wrapper_volume_name = format!("{}-zombie-wrapper", self.name);
+
+        self.docker_client
+            .create_volume(&zombie_wrapper_volume_name)
+            .await
+            .map_err(|err| ProviderError::CreateNamespaceFailed(self.name.clone(), err.into()))?;
+
+        // copy script to volume
+        self.docker_client
+            .container_run(
+                ContainerRunOptions::new(
+                    "alpine:latest",
+                    vec!["cp", "/zombie-wrapper.sh", "/scripts"],
```

Comment:
Can we just add `"&&", "chmod", "-R", "777", "/scripts"` to _not_ need to create a new container for just this op.